llama.cpp

```diff
@@ -15697,6 +15697,58 @@ static int32_t llama_chat_apply_template_internal(
                 ss << message->content << "</s>";
             }
         }
+    } else if (tmpl == "openchat" || tmpl.find("GPT4 Correct ") != std::string::npos) {
+        // openchat/openchat-3.5-0106,
+        for (auto message : chat) {
+            std::string role(message->role);
+            if (role == "user") {
+                ss << "GPT4 Correct User: ";
+            } else if (role == "assistant") {
+                ss << "GPT4 Correct Assistant: ";
+            }
+            // Not documented, but apparently the system message is prepended without prefix:
+            // https://huggingface.co/openchat/openchat_3.5/discussions/5#65448109b4a3f3a2f486fd9d
+            ss << message->content << "<|end_of_turn|>";
+        }
+        if (add_ass) {
+            ss << "GPT4 Correct Assistant:";
+        }
+    } else if (tmpl == "vicuna" || tmpl == "vicuna-orca" || (tmpl.find("USER: ") != std::string::npos && tmpl.find("ASSISTANT: ") != std::string::npos)) {
+        // eachadea/vicuna-13b-1.1 (and Orca variant)
+        for (auto message : chat) {
+            std::string role(message->role);
+            if (role == "system") {
+                // Orca-Vicuna variant uses a system prefix
+                if (tmpl == "vicuna-orca" || tmpl.find("SYSTEM: ") != std::string::npos) {
+                    ss << "SYSTEM: " << message->content << "\n";
+                } else {
+                    ss << message->content << "\n\n";
+                }
+            } else if (role == "user") {
+                ss << "USER: " << message->content << "\n";
+            } else if (role == "assistant") {
+                ss << "ASSISTANT: " << message->content << "</s>\n";
+            }
+        }
+        if (add_ass) {
+            ss << "ASSISTANT:";
+        }
+    } else if (tmpl == "alpaca" || tmpl.find("### Instruction:\\n") != std::string::npos) {
+        // wxjiao/alpaca-7b, deepseek-ai/deepseek-coder-33b-instruct
+        for (auto message : chat) {
+            std::string role(message->role);
+            if (role == "system") {
+                ss << message->content << "\n\n";
+            } else if (role == "user") {
+                ss << "### Instruction:\n" << message->content << "\n\n";
+            } else if (role == "assistant") {
+                ss << "### Response:\n" << message->content << "\n\n";
```

Comment:
alpaca template and deepseek template both look similar at the first glance, but the main different is that alpaca template only used for instruction-response (one turn) and not multiple turns like modern chat template.

deepseek extends the notion of instruction-response into multi-turn by placing `<|EOT|>` token between each turn, so the formatted chat should look like:

```
### Instruction:
who are you?
### Response:
I am assistant
<|EOT|>
### Instruction:
1+1 is
### Response:
equal to 2
<|EOT|>
```

So what missing here is that `<|EOT|>`

Reply:
The chat above is produced by python code + jinja template, it doesn't seem to have `"\n\n"` at the end of each message, so I think the `"\n\n"` should be replaced by `"\n"`

Reply:
Thanks for the python script! Included the Jinja output of OpenChat and DeepSeek below. And as you mentioned, the other two fail due to not having templates in `config_tokenizers.json`.

Will add `<|EOT|>` for DeepSeek when I have moment tomorrow.

```
openchat/openchat-3.5-0106
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
<s>GPT4 Correct System: You are a helpful assistant<|end_of_turn|>GPT4 Correct User: Hello<|end_of_turn|>GPT4 Correct Assistant: Hi there<|end_of_turn|>GPT4 Correct User: Who are you<|end_of_turn|>GPT4 Correct Assistant:    I am an assistant   <|end_of_turn|>GPT4 Correct User: Another question<|end_of_turn|>
------------------------------
deepseek-ai/deepseek-coder-33b-instruct
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
<｜begin▁of▁sentence｜>You are a helpful assistant### Instruction:
Hello
### Response:
Hi there
<|EOT|>
### Instruction:
Who are you
### Response:
   I am an assistant   
<|EOT|>
### Instruction:
Another question

------------------------------
```

Reply:
Looks OK to me, just a quick note is that `<｜begin▁of▁sentence｜>` is not needed, because BOS is always added on server